Mathlib/Algebra/Star/NonUnitalSubalgebra.lean

```diff
@@ -0,0 +1,1224 @@
+#align non_unital_subalgebra.star_mono NonUnitalSubalgebra.star_mono
+
+variable (R)
+
+/-- The star operation on `NonUnitalSubalgebra` commutes with `NonUnitalAlgebra.adjoin`. -/
+theorem star_adjoin_comm (s : Set A) :
+    star (NonUnitalAlgebra.adjoin R s) = NonUnitalAlgebra.adjoin R (star s) :=
+  have this :
+    ∀ t : Set A, NonUnitalAlgebra.adjoin R (star t) ≤ star (NonUnitalAlgebra.adjoin R t) := fun t =>
+    NonUnitalAlgebra.adjoin_le fun x hx => NonUnitalAlgebra.subset_adjoin R hx
+  le_antisymm (by simpa only [star_star] using NonUnitalSubalgebra.star_mono (this (star s)))
+    (this s)
+#align non_unital_subalgebra.star_adjoin_comm NonUnitalSubalgebra.star_adjoin_comm
+
+variable {R}
+
+/-- The `NonUnitalStarSubalgebra` obtained from `S : NonUnitalSubalgebra R A` by taking the
+smallest non-unital subalgebra containing both `S` and `star S`. -/
+@[simps!]
+def starClosure (S : NonUnitalSubalgebra R A) : NonUnitalStarSubalgebra R A where
+  toNonUnitalSubalgebra := S ⊔ star S
+  star_mem' := @fun a (ha : a ∈ S ⊔ star S) => show star a ∈ S ⊔ star S by
+    simp only [← mem_star_iff _ a, ← (@NonUnitalAlgebra.gi R A _ _ _ _ _).l_sup_u _ _] at *
+    convert ha using 2
+    simp only [Set.sup_eq_union, star_adjoin_comm, Set.union_star, coe_star, star_star,
+      Set.union_comm]
+#align non_unital_subalgebra.star_closure NonUnitalSubalgebra.starClosure
+
+theorem starClosure_le {S₁ : NonUnitalSubalgebra R A} {S₂ : NonUnitalStarSubalgebra R A}
+    (h : S₁ ≤ S₂.toNonUnitalSubalgebra) : S₁.starClosure ≤ S₂ :=
+  NonUnitalStarSubalgebra.toNonUnitalSubalgebra_le_iff.1 <|
+    sup_le h fun x hx =>
+      (star_star x ▸ star_mem (show star x ∈ S₂ from h <| (S₁.mem_star_iff _).1 hx) : x ∈ S₂)
+#align non_unital_subalgebra.star_closure_le NonUnitalSubalgebra.starClosure_le
+
+theorem starClosure_le_iff {S₁ : NonUnitalSubalgebra R A} {S₂ : NonUnitalStarSubalgebra R A} :
+    S₁.starClosure ≤ S₂ ↔ S₁ ≤ S₂.toNonUnitalSubalgebra :=
+  ⟨fun h => le_sup_left.trans h, starClosure_le⟩
+#align non_unital_subalgebra.star_closure_le_iff NonUnitalSubalgebra.starClosure_le_iff
+
+@[simp]
+theorem starClosure_toNonunitalSubalgebra {S : NonUnitalSubalgebra R A} :
+    S.starClosure.toNonUnitalSubalgebra = S ⊔ star S :=
+  rfl
+
+@[mono]
+theorem starClosure_mono : Monotone (starClosure (R := R) (A := A)) :=
+  fun _ _ h => starClosure_le <| h.trans le_sup_left
+
+end NonUnitalSubalgebra
+
+namespace NonUnitalStarAlgebra
+
+variable [CommSemiring R] [StarRing R]
+variable [NonUnitalSemiring A] [StarRing A]
+variable [Module R A] [IsScalarTower R A A] [SMulCommClass R A A] [StarModule R A]
+variable [NonUnitalSemiring B] [StarRing B]
+variable [Module R B] [IsScalarTower R B B] [SMulCommClass R B B] [StarModule R B]
+variable [hF : NonUnitalStarAlgHomClass F R A B]
+
+open scoped Pointwise
+
+open NonUnitalStarSubalgebra
+
+variable (R)
+
+/-- The minimal non-unital subalgebra that includes `s`. -/
+def adjoin (s : Set A) : NonUnitalStarSubalgebra R A where
```

Comment:
I presume `adjoin` is standard terminology here? I note that `NonUnitalAlgebra.adjoin` already exists.

(I would have expected `closure` or similar.)